addons/l10n_ke_edi_tremol/models/account_move.py

```diff
@@ -0,0 +1,227 @@
+                if line_to_discount >= still_to_discount:
+                    discount_dict[candidate.id] = 100.0
+                    line_to_discount -= still_to_discount
+                else:
+                    rest_to_discount = abs((line_to_discount / (candidate.price_unit * candidate.quantity)) * 100.0)
+                    discount_dict[candidate.id] += rest_to_discount
+                    break
+
+        vat_class = {16.0: 'A', 8.0: 'B'}
+        msgs = []
+        for line in self.invoice_line_ids.filtered(lambda l: not l.display_type and l.quantity and l.price_total > 0 and not discount_dict.get(l.id) >= 100):
+            # Here we use the original discount of the line, since it the distributed discount has not been applied in the price_total
+            price = round(line.price_total / line.quantity * 100 / (100-line.discount), 2)
+            percentage = line.tax_ids[0].amount
+
+            # Letter to classify tax, 0% taxes are handled conditionally, as the tax can be zero-rated or exempt
+            letter = ''
+            if percentage in vat_class:
+                letter = vat_class[percentage]
+            else:
+                report_line_ids = line.tax_ids[0].invoice_repartition_line_ids.filtered(lambda r: r.repartition_type == 'base').tag_ids.tax_report_line_ids.ids
+                letter = 'E' if self.env.ref('l10n_ke.tax_report_line_exempt_sales_base', raise_if_not_found=False).id in report_line_ids else 'C'
+
+            uom = line.product_uom_id and line.product_uom_id.name or ''
+            if letter != 'A':
+                hscode = line.product_id.l10n_ke_hsn_code
+                hsname = line.product_id.l10n_ke_hsn_name
+            else:
+                hscode = ''
+                hsname = ''
+            line_data = b';'.join([
+                self._l10n_ke_fmt(line.name, 36),               # 36 symbols for the article's name
+                self._l10n_ke_fmt(letter, 1),                   # 1 symbol for article's vat class ('A', 'B', 'C', 'D', or 'E')
+                str(price)[:13].encode('cp1251'),                # 1 to 13 symbols for article's price
+                self._l10n_ke_fmt(uom, 3),                      # 3 symbols for unit of measure
+                self._l10n_ke_fmt(hscode, 10),                  # 10 symbols for HS code in the format xxxx.xx.xx (can be empty)
+                self._l10n_ke_fmt(hsname, 20),                  # 20 symbols for the HS name (can be empty)
+                str(percentage).encode('cp1251')[:5]             # up to 5 symbols for vat rate
+            ])
+
+            # 1 to 10 symbols for quantity
+            line_data += b'*' + str(line.quantity).encode('cp1251')[:10]
+            if discount_dict[line.id]:
+                # 1 to 7 symbols for percentage of discount/addition
+                discount_sign = b'-' if discount_dict[line.id] > 0 else b'+'
```

Comment:
You confirm that if discount 0.0, we do not need to add anything?

Reply:
![image](https://user-images.githubusercontent.com/46533567/204001524-a1a3fd87-c1bc-4f48-b64e-263d17893307.png)

yes it's optional